Firestore/core/src/firebase/firestore/api/query_core.mm

```diff
@@ -0,0 +1,327 @@
+      }
+      ResourcePath path =
+          query().path.Append(ResourcePath::FromString(document_key));
+      if (!DocumentKey::IsDocumentKey(path)) {
+        ThrowInvalidArgument(
+            "Invalid query. When querying a collection group by document ID, "
+            "the value provided must result in a valid document path, but '%s' "
+            "is not because it has an odd number of segments.",
+            path.CanonicalString());
+      }
+      field_value = [FSTReferenceValue
+          referenceValue:[FSTDocumentKey keyWithDocumentKey:DocumentKey{path}]
+              databaseID:&firestore_->database_id()];
+    } else if (field_value.type != FieldValue::Type::Reference) {
+      ThrowInvalidArgument(
+          "Invalid query. When querying by document ID you must provide a "
+          "valid string or DocumentReference, but it was of type: %s",
+          type_describer());
+    }
+  }
+
+  FSTFilter* filter = [FSTFilter filterWithField:field_path
+                                  filterOperator:op
+                                           value:field_value];
+
+  if ([filter isKindOfClass:[FSTRelationFilter class]]) {
+    ValidateNewRelationFilter(static_cast<FSTRelationFilter*>(filter));
+  }
+
+  return Wrap([query_ queryByAddingFilter:filter]);
+}
+
+Query Query::OrderBy(FieldPath fieldPath, bool descending) const {
+  return OrderBy(fieldPath, Direction::FromDescending(descending));
+}
+
+Query Query::OrderBy(FieldPath fieldPath, Direction direction) const {
+  ValidateNewOrderByPath(fieldPath);
+  if (query().startAt) {
+    ThrowInvalidArgument("Invalid query. You must not specify a starting point "
+                         "before specifying the order by.");
+  }
+  if (query().endAt) {
+    ThrowInvalidArgument("Invalid query. You must not specify an ending point "
+                         "before specifying the order by.");
+  }
+  bool ascending = direction == Direction::Ascending;
+  FSTSortOrder* sortOrder = [FSTSortOrder sortOrderWithFieldPath:fieldPath
+                                                       ascending:ascending];
+  return Wrap([query() queryByAddingSortOrder:sortOrder]);
+}
+
+Query Query::Limit(int64_t limit) const {
+  if (limit <= 0) {
+    ThrowInvalidArgument(
+        "Invalid Query. Query limit (%s) is invalid. Limit must be positive.",
+        limit);
+  }
+  return Wrap([query() queryBySettingLimit:limit]);
+}
+
+Query Query::StartAt(FSTBound* bound) const {
+  return Wrap([query() queryByAddingStartAt:bound]);
+}
+
+Query Query::EndAt(FSTBound* bound) const {
+  return Wrap([query() queryByAddingEndAt:bound]);
+}
+
+void Query::ValidateNewRelationFilter(FSTRelationFilter* filter) const {
+  if ([filter isInequality]) {
+    const FieldPath* existingField = [query_ inequalityFilterField];
+    if (existingField && *existingField != filter.field) {
+      ThrowInvalidArgument(
+          "Invalid Query. All where filters with an inequality (lessThan, "
```

Comment:
Optional: is it possible to emphasize `where` so that it's clear that it's a "keyword"? E.g. by putting it into single quotes, etc.

Reply:
Done.